core/src/main/java/com/microsoft/applicationinsights/internal/perfcounter/OshiPerformanceCounter.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * ApplicationInsights-Java
+ * Copyright (c) Microsoft Corporation
+ * All rights reserved.
+ *
+ * MIT License
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of this
+ * software and associated documentation files (the ""Software""), to deal in the Software
+ * without restriction, including without limitation the rights to use, copy, modify, merge,
+ * publish, distribute, sublicense, and/or sell copies of the Software, and to permit
+ * persons to whom the Software is furnished to do so, subject to the following conditions:
+ * The above copyright notice and this permission notice shall be included in all copies or
+ * substantial portions of the Software.
+ * THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
+ * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR
+ * PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
+ * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR
+ * OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ */
+
+package com.microsoft.applicationinsights.internal.perfcounter;
+
+import com.microsoft.applicationinsights.TelemetryClient;
+import com.microsoft.applicationinsights.telemetry.MetricTelemetry;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import oshi.SystemInfo;
+import oshi.hardware.HWDiskStore;
+import oshi.hardware.HardwareAbstractionLayer;
+import oshi.software.os.OSProcess;
+import oshi.software.os.OperatingSystem;
+
+public class OshiPerformanceCounter implements PerformanceCounter {
+
+    private static final Logger logger = LoggerFactory.getLogger(OshiPerformanceCounter.class);
+    private static final String ID = Constants.PERFORMANCE_COUNTER_PREFIX + "OshiPerformanceCounter";
+    private final static double NANOS_IN_SECOND = 1_000_000_000;
+    private long lastCollectionInNanos = -1;
+    private double prevProcessIO;
+    private HardwareAbstractionLayer hal;
+    private OSProcess processInfo;
+
+    public OshiPerformanceCounter() {
+        SystemInfo systemInfo = new SystemInfo();
+        hal = systemInfo.getHardware();
+        OperatingSystem osInfo = systemInfo.getOperatingSystem();
+        processInfo = osInfo.getProcess(osInfo.getProcessId());
+    }
+
+    @Override public String getId() {
+        return ID;
+    }
+
+    @Override public void report(TelemetryClient telemetryClient) {
+        long currentCollectionInNanos = System.nanoTime();
+        double currentProcessIO = 0.0;
+
+        // process io
+        currentProcessIO += (double) (processInfo.getBytesRead() + processInfo.getBytesWritten());
+
+        if (lastCollectionInNanos != -1) {
```

Comment:
weirdly, System.nanoTime() can return negative values, so would probably be better to track the first time in a different way (maybe just a new field), just in the slim slim case that it returns -1

Reply:
System.nanoTime() won't return negative values.. -1 is the initialization to make sure it skips this if block for the very 1st time

Reply:
check out https://docs.oracle.com/javase/8/docs/api/java/lang/System.html#nanoTime--